### PR TITLE
Publish `pyk` on PyPI

### LINKED ISSUE #4239: Publish `pyk` package
Since the K repository has a lot of submodules, installing `pyk` from a `git` dependency using Poetry is very slow.

We should push a distribution package to PyPI. Alternatively, we can publish the source archive as a GitHub release.

### PR DESCRIPTION
**Pull request by F-WRunTime**
Fixes #4239

~Blocked on~:
* #4438
* #4461

Add job `pyk-publish` to publish `pyk` on PyPI (under package name `kframework`).

Release workflow changes from

```
       release --- notify-dependents
                \
pyk-build-docs --- gh-pages
```

to

```
release --- pyk-publish --- notify-dependents
                         \
         pyk-build-docs --- gh-pages

```

### CODE REVIEW

File: pyk/src/kframework/__init__.py
Comment (tothtamas28) [inline pyk/src/kframework/__init__.py:1]: The `pyk` package should not be renamed, it can still be distributed under `kframework`.

But since the module name is no longer the default one, `pyproject.toml` should be updated (not tested):

```toml
packages = [
    { include = "pyk", from = "src" },
]
```

To not mess up the template, `cruft.json` (and everything that depends on the corresponding field) should be updated.

Later, if we want to rename the package, we can do so without immediately breaking dependents by distributing it under both names during the transition.

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml]: Renaming should be the only thing this PR takes care of.

File: pyk/src/kframework/__init__.py
Comment (tothtamas28) [reply pyk/src/kframework/__init__.py:1]: Even if only the distribution is renamed, this change is breaking to all dependents. I think the change necessary is

```toml
[dependencies]
pyk = {}
```

to

```toml
[dependencies]
kframework = {}
```

in `pyproject.toml` (and `poetry.lock`, and the CI jobs that depend on it, `update.yml` in particular). In any case, before merging this, it should be tested on `evm-semantics`.

**Review by tothtamas28 — DISMISSED** [review]
The base of the PR should be `develop`.

**Comment by Baltoli** [comment]
I agree with Tamas here - let's make the minimum renaming change that gets us to a package that could be published, then do the pypi steps in a follow-up PR.

**Comment by F-WRunTime** [comment]
> The base of the PR should be `develop`.

Unclear why the master-pr workflow didn't change it. Thanks for pointing it out.

File: pyk/src/kframework/__init__.py
Comment (F-WRunTime) [reply pyk/src/kframework/__init__.py:1]: I'm not sure I follow, this is an issue I brought up on slack. The pyk package name is already taken on Pypi and we cannot upload packages with the same name they must be unique. 

> ```toml
> ```toml
> packages = [
>     { include = "pyk", from = "src" },
> ]
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```

I don't see this definition in the src. Which file are you referring to?

Comment (tothtamas28) [reply pyk/src/kframework/__init__.py:1]: > The pyk package name is already taken on Pypi

In Python packaging terminology, the term _package_ is overloaded, as it means two different things:
- An _import package_:
  ```python
  import pyk
  ```
- A _distribution package_:
  ```bash
  $ pip install pyk
  ```

A distribution package can ship more than one import packages, whose names are arbitrary well-formed import package names.

PyPI expects the name of the distribution package to be unique. I haven't tried this, but my presumption is, it does not put any restriction on the names of import packages. So as long as we do not want to install two distribution packages that each ship an import package with the name `pyk`, we're fine.

Comment (tothtamas28) [reply pyk/src/kframework/__init__.py:1]: > I don't see this definition in the src. Which file are you referring to?

`pyproject.toml`. It's not there, as by default Poetry looks for an import package with the same name as the distribution package. If you want to change that, you need to include the `packages` key.

Comment (tothtamas28) [reply pyk/src/kframework/__init__.py:1]: Btw, I'm happy to do these changes. Then you can review them and try to upload the resulting (distribution) package to the test repository.

**Comment by Baltoli** [comment]
This is partially superseded by https://github.com/runtimeverification/k/issues/4438, right? I presume after @tothtamas28's PR is merged that this one should be refactored to do _only_ the push to PyPi.

**Comment by F-WRunTime** [comment]
> This is partially superseded by #4438, right? I presume after @tothtamas28's PR is merged that this one should be refactored to do _only_ the push to PyPi.

@Baltoli Correct. Waiting on merge from the other PR first.

**Review by Baltoli — DISMISSED** [review]

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml]:
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml]: I don't think the action supports that field: https://github.com/Gr1N/setup-poetry

```suggestion
```

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml]:
```suggestion
        uses: actions/setup-python@v5
```

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml]: There's not much to cache, `poetry build` does not require a virtual environment.

```suggestion
```

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml]:
```suggestion
      - name: Publish pyk to PyPI
```

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml]:
```suggestion
        uses: pypa/gh-action-pypi-publish@v1
```

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml]: Trailing newline got removed here.

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml]: Nitpick:

```suggestion
      - name: Install Python
```

For consistency with the next step and https://github.com/runtimeverification/k/blob/f373630f28653b60b4079cce2242e3bab90ef42c/.github/actions/setup-pyk-env/action.yml#L12

**Comment by tothtamas28** [comment]
When should this step occur? The way I see it, constraints are:
- Should happen after K is published: K is an implicit dependency of `pyk`.
- Should happen after `pyk` documentation is built: the published package refers to the API docs.
- Should happen before notifying dependents: dependency update PRs will refer to the package.

So maybe we should have:
```
       release --- pyk-publish - gh-pages - notify-dependents
pyk-build-docs -/
```

Where `notify-dependents` is a new step extracted from `release`.

@Baltoli, @F-WRunTime, what do you think?

**Comment by Baltoli** [comment]
> When should this step occur?

I agree with your assessment; you've captured the relevant functional dependencies there

**Review by Baltoli — APPROVED** [review]

File: .github/workflows/release.yml
Comment (tothtamas28) [inline .github/workflows/release.yml:412]: I wonder, is this part necessary? @Baltoli @F-WRunTime

Comment (Baltoli) [reply .github/workflows/release.yml:412]: Unsure; I don't _think_ so looking at the docs but I can't say authoritatively

Comment (tothtamas28) [reply .github/workflows/release.yml:412]: The secret is probably defined for the `production` environment.

The other part is apparently for signing the published package: https://github.com/pypa/gh-action-pypi-publish/blob/unstable/v1/README.md?plain=1#L63-L64

Comment (Baltoli) [reply .github/workflows/release.yml:412]: Makes sense!

Comment (F-WRunTime) [reply .github/workflows/release.yml:412]: Sorry seeing this now, but yes it is necessary. I want to transition to using a verified publisher approach this sets us up for that. And yes the secret is stored in that environment. It's ultimately a better practice than storing everything in a general environment.

**Pull request merged**
